pylops/utils/decorators.py

```diff
@@ -121,3 +121,42 @@ def wrapper(self, x):
     if func is not None:
         return decorator(func)
     return decorator
+
+
+def count(f=None, forward=True, mat=False):
+    """Decorator used to count the number of forward and adjoint performed by an operator.
+
+    Parameters
+    ----------
+    f : :obj:`callable`, optional
+        Function to be decorated when no arguments are provided
+    forward : :obj:`bool`, optional
+        Whether to count the forward (``True``) or adjoint (``False``)
+    mat : :obj:`bool`, optional
+        Whether to count the matvec (``False``) or matmat (``True``)
+
+    """
+
+    def decorator(f):
+        @wraps(f)
+        def wrapper(self, x):
+            if forward:
+                if mat:
+                    self.matmat_count += 1
+                    self.matvec_count -= x.shape[-1]
+                else:
+                    self.matvec_count += 1
+            else:
+                if mat:
+                    self.rmatmat_count += 1
+                    self.rmatvec_count -= x.shape[-1]
+                else:
+                    self.rmatvec_count += 1
+            y = f(self, x)
```

Comment:
I think it's better to put this line in the beginning of the wrapper. If for some reason `f` errors, the linear operator will have the wrong count.

Reply:
Agree :)